tests/core/test_hook_function_single_gpu.py

```diff
@@ -0,0 +1,137 @@
+from flex_model.core import FlexModel, HookFunction
+import torch
+import torch.nn as nn
+from typing import Dict
+from argparse import Namespace
+from transformers import AutoModelForCausalLM, AutoTokenizer
+from functools import partial
+from tests.registry import register_test
+
+
+# could be any MLP layer and the code won't break. The test doesn't generalize
+# to other kinds of layers
+MODULE_NAME = "model.layers.27.mlp"
+
+
+class LayerInjection(nn.Module):
+    """
+    Purpose is to replace an MLP module in the model. In short, this class
+    allows us to store and modify an incoming activation without using any
+    sort of forward hooks. It acts as the absolute ground truth for later
+    comparison.
+    """
+
+    def __init__(self, module_name: str, model: nn.Module):
+        super(LayerInjection, self).__init__()
+        self.original_out: torch.Tensor = None
+        self.modified_out: torch.Tensor = None
+        self.decoder_layer: nn.Module = None
+
+        for n, m in model.named_modules():
+            if n == module_name:
+                self.decoder_layer = m
+
+    def __call__(self, *args, **kwargs) -> torch.Tensor:
+        out = self.decoder_layer(*args, **kwargs)
+        self.original_out = out.detach()
+        self.modified_out = self.original_out * 2
+        return out
+
+
+def inject_module(injection_layer: nn.Module, model: nn.Module) -> None:
+    """Replace <MODULE_NAME> in <model> with <injection_layer>"""
+    sliced = MODULE_NAME.split(".")
+    layer_num = int(sliced[2])
+    model.model.layers[layer_num].mlp = injection_layer
+
+
+def editing_func(
+    current_module: nn.Module,
+    inputs: torch.Tensor,
+    save_ctx: Namespace,
+    modules: nn.ModuleDict,
+    in_dict: Dict[str, torch.Tensor],
+) -> torch.Tensor:
+    """
+    Passed into HookFunction. Takes the input tensor, multiplies it by 2, saves it and
+    returns the input tensor.
+    """
+    multiplied = inputs * 2
+    in_dict["modified"] = multiplied.detach().cpu()
+    return inputs
+
+
+def make_model_and_tokenizer():
```

Comment:
Consider pulling the model and tokenizer construction functions into the `tests.test_utilities.py` file and importing.